Vim 8.1.800+ repeatable crash/lockup due to `heap-use-after-free`

Vim 8.1.800 and later exhibit crashes and lockups under repeatable conditions explained below.  Patch 8.1.800 introduced code to invoke `garbage_collect` every 10,000 times a copy is made.  Under certain patterns of Vimscript usage, this leads to crashes.

The bug can be triggered using some test code unknowingly written by Christian Brabant (that is, a particular version of his vim-airline plugin triggers the crashing).

# To reproduce the crash

- Prepare a suitable Linux environment (I'm using Ubuntu 18.04 on x86_64, but presumably others would work as well).

- Setup a clean user environment with the particular version of vim-airline as the only content:

```sh
mv ~/.vimrc ~/orig.vimrc
mv ~/.vim ~/orig.vim
git clone https://github.com/vim-airline/vim-airline ~/.vim
cd ~/.vim
git checkout c7fb175
```

- Checkout the most recent Vim source (tested in this ticket using 8.1.0867):

```sh
git clone https://github.com/vim/vim ~/vim
cd ~/vim
```

- Apply the following patch in the Makefile to turn on the address sanitizer:

```diff
diff --git a/src/Makefile b/src/Makefile
index 3f5418456..639f6370b 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -690,7 +690,7 @@ LINT_OPTIONS = -beprxzF
 # address sanitizer or with the undefined sanitizer.  Works with gcc and
 # clang.  May make Vim twice as slow.  Errors reported on stderr.
 # More at: https://code.google.com/p/address-sanitizer/
-#SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
+SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
 #SANITIZER_CFLAGS = -g -O0 -fsanitize=undefined -fno-omit-frame-pointer
 SANITIZER_LIBS = $(SANITIZER_CFLAGS)
```

- Apply the following patch to make the garbage collector run after every copy instead of every 10,000 copies to trigger the bug more rapidly:

```diff
diff --git a/src/userfunc.c b/src/userfunc.c
index aae3fd334..a54361c42 100644
--- a/src/userfunc.c
+++ b/src/userfunc.c
@@ -675,7 +675,7 @@ cleanup_function_call(funccall_T *fc)
        for (li = fc->l_varlist.lv_first; li != NULL; li = li->li_next)
            copy_tv(&li->li_tv, &li->li_tv);

-	if (++made_copy == 10000)
+	if (++made_copy == 1)
        {
            // We have made a lot of copies.  This can happen when
            // repetitively calling a function that creates a reference to
```

- Configure and build Vim:

```sh
./configure
make
```

- Invoke Vim; it should abort during startup and write debugging information to stderr.  Reset the terminal afterward (necessary on my machine, at least):

```sh
./src/vim 2> crash.txt; reset; cat crash.txt
```

- Examine the crash output; it should look similar to sample shown at the bottom of this ticket.

I haven't tried to figure out what is special about the version of vim-airline that allows it to trigger the bug.  The HEAD revision of vim-airline doesn't seem to crash (at least, not at Vim startup).

# Sample crash.txt output

```
=================================================================
==27891==ERROR: AddressSanitizer: heap-use-after-free on address 0x60700001ea58 at pc 0x55cc0fe49b0f bp 0x7ffc019565e0 sp 0x7ffc019565d0
READ of size 4 at 0x60700001ea58 thread T0
    #0 0x55cc0fe49b0e in copy_tv /home/drmikehenry/vim/src/eval.c:8140
    #1 0x55cc0fe48cbc in set_var /home/drmikehenry/vim/src/eval.c:7947
    #2 0x55cc0fe2d20b in set_var_lval /home/drmikehenry/vim/src/eval.c:2343
    #3 0x55cc0fe298d4 in ex_let_one /home/drmikehenry/vim/src/eval.c:1825
    #4 0x55cc0fe26e68 in ex_let_vars /home/drmikehenry/vim/src/eval.c:1304
    #5 0x55cc0fe2f86a in next_for_item /home/drmikehenry/vim/src/eval.c:2690
    #6 0x55cc0ff232ef in ex_while /home/drmikehenry/vim/src/ex_eval.c:1097
    #7 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #8 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #9 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #10 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #11 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #12 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #13 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #14 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #15 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #16 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #17 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #18 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #19 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #20 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #21 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #22 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #23 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #24 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #25 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #26 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #27 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #28 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #29 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #30 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #31 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #32 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #33 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #34 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #35 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #36 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #37 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #38 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #39 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #40 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #41 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #42 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #43 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #44 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #45 0x55cc0fdaea40 in apply_autocmds_group /home/drmikehenry/vim/src/autocmd.c:2072
    #46 0x55cc0fdad3a6 in apply_autocmds_retval /home/drmikehenry/vim/src/autocmd.c:1622
    #47 0x55cc0fdb51a4 in open_buffer /home/drmikehenry/vim/src/buffer.c:332
    #48 0x55cc103f8577 in create_windows /home/drmikehenry/vim/src/main.c:2746
    #49 0x55cc103f172b in vim_main2 /home/drmikehenry/vim/src/main.c:736
    #50 0x55cc103f0eeb in main /home/drmikehenry/vim/src/main.c:441
    #51 0x7fa5e7e72b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #52 0x55cc0fda5e79 in _start (/home/drmikehenry/vim/src/vim+0x111e79)

0x60700001ea58 is located 56 bytes inside of 80-byte region [0x60700001ea20,0x60700001ea70)
freed by thread T0 here:
    #0 0x7fa5eb1577b8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xde7b8)
    #1 0x55cc10030126 in vim_free /home/drmikehenry/vim/src/misc2.c:1817
    #2 0x55cc0ffd4772 in list_free_list /home/drmikehenry/vim/src/list.c:208
    #3 0x55cc0ffd484b in list_free_items /home/drmikehenry/vim/src/list.c:225
    #4 0x55cc0fe3d5fd in free_unref_items /home/drmikehenry/vim/src/eval.c:5503
    #5 0x55cc0fe3d4f8 in garbage_collect /home/drmikehenry/vim/src/eval.c:5449
    #6 0x55cc10305673 in cleanup_function_call /home/drmikehenry/vim/src/userfunc.c:685
    #7 0x55cc10307e72 in call_user_func /home/drmikehenry/vim/src/userfunc.c:1066
    #8 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #9 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #10 0x55cc0fe37926 in eval7 /home/drmikehenry/vim/src/eval.c:4382
    #11 0x55cc0fe35e64 in eval6 /home/drmikehenry/vim/src/eval.c:3985
    #12 0x55cc0fe35017 in eval5 /home/drmikehenry/vim/src/eval.c:3781
    #13 0x55cc0fe34773 in eval4 /home/drmikehenry/vim/src/eval.c:3665
    #14 0x55cc0fe34307 in eval3 /home/drmikehenry/vim/src/eval.c:3585
    #15 0x55cc0fe33e65 in eval2 /home/drmikehenry/vim/src/eval.c:3517
    #16 0x55cc0fe338ee in eval1 /home/drmikehenry/vim/src/eval.c:3445
    #17 0x55cc0fe335f3 in eval0 /home/drmikehenry/vim/src/eval.c:3403
    #18 0x55cc0fe26afe in ex_let /home/drmikehenry/vim/src/eval.c:1260
    #19 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #20 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #21 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #22 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #23 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #24 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #25 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #26 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #27 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #28 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #29 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455

previously allocated by thread T0 here:
    #0 0x7fa5eb157b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x55cc1002ddce in lalloc /home/drmikehenry/vim/src/misc2.c:944
    #2 0x55cc1002dc41 in alloc_clear /home/drmikehenry/vim/src/misc2.c:853
    #3 0x55cc0ffd40b1 in list_alloc /home/drmikehenry/vim/src/list.c:75
    #4 0x55cc0fddd6ae in dict_list /home/drmikehenry/vim/src/dict.c:875
    #5 0x55cc0fe765ad in f_items /home/drmikehenry/vim/src/evalfunc.c:7393
    #6 0x55cc0fe58535 in call_internal_func /home/drmikehenry/vim/src/evalfunc.c:1117
    #7 0x55cc10309b7f in call_func /home/drmikehenry/vim/src/userfunc.c:1520
    #8 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #9 0x55cc0fe37926 in eval7 /home/drmikehenry/vim/src/eval.c:4382
    #10 0x55cc0fe35e64 in eval6 /home/drmikehenry/vim/src/eval.c:3985
    #11 0x55cc0fe35017 in eval5 /home/drmikehenry/vim/src/eval.c:3781
    #12 0x55cc0fe34773 in eval4 /home/drmikehenry/vim/src/eval.c:3665
    #13 0x55cc0fe34307 in eval3 /home/drmikehenry/vim/src/eval.c:3585
    #14 0x55cc0fe33e65 in eval2 /home/drmikehenry/vim/src/eval.c:3517
    #15 0x55cc0fe338ee in eval1 /home/drmikehenry/vim/src/eval.c:3445
    #16 0x55cc0fe335f3 in eval0 /home/drmikehenry/vim/src/eval.c:3403
    #17 0x55cc0fe2efd6 in eval_for_line /home/drmikehenry/vim/src/eval.c:2608
    #18 0x55cc0ff231c7 in ex_while /home/drmikehenry/vim/src/ex_eval.c:1091
    #19 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #20 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #21 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #22 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #23 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455
    #24 0x55cc10314649 in ex_call /home/drmikehenry/vim/src/userfunc.c:3184
    #25 0x55cc0fee7f37 in do_one_cmd /home/drmikehenry/vim/src/ex_docmd.c:2520
    #26 0x55cc0fedf44f in do_cmdline /home/drmikehenry/vim/src/ex_docmd.c:1033
    #27 0x55cc1030719a in call_user_func /home/drmikehenry/vim/src/userfunc.c:967
    #28 0x55cc10309a27 in call_func /home/drmikehenry/vim/src/userfunc.c:1501
    #29 0x55cc10304829 in get_func_tv /home/drmikehenry/vim/src/userfunc.c:455

SUMMARY: AddressSanitizer: heap-use-after-free /home/drmikehenry/vim/src/eval.c:8140 in copy_tv
Shadow bytes around the buggy address:
  0x0c0e7fffbcf0: fd fd fd fd fd fd fd fd fd fa fa fa fa fa 00 00
  0x0c0e7fffbd00: 00 00 00 00 00 00 02 fa fa fa fa fa fd fd fd fd
  0x0c0e7fffbd10: fd fd fd fd fd fa fa fa fa fa 00 00 00 00 00 00
  0x0c0e7fffbd20: 00 00 02 fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x0c0e7fffbd30: 00 00 fa fa fa fa 00 00 00 00 00 00 00 00 00 00
=>0x0c0e7fffbd40: fa fa fa fa fd fd fd fd fd fd fd[fd]fd fd fa fa
  0x0c0e7fffbd50: fa fa fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0e7fffbd60: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fd fd
  0x0c0e7fffbd70: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0e7fffbd80: fd fd fd fd fd fd fa fa fa fa fd fd fd fd fd fd
  0x0c0e7fffbd90: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==27891==ABORTING
```

I can reproduce a crash in the latest vim (8.1.0867) after patching
`userfunc.c` as suggested by @drmikehenry  i.e. replacing
`if (++made_copy == 10000)` with  `if (++made_copy == 1)`
and using a asan build, but without even using vim-airline, and
instead run a vim test:

```
$ cd vim/src/testdir
$ make test_lambda
```
Commenting out all tests in test_lambda.vim, I see that 
`Test_lambda_side_effect()` alone is enough to trigger
an asan error:

```
=================================================================
==10998==ERROR: AddressSanitizer: heap-use-after-free on address 0x6070000038c8 at pc 0x564841733ca6 bp 0x7ffe969b8820 sp 0x7ffe969b8810
READ of size 4 at 0x6070000038c8 thread T0
    #0 0x564841733ca5 in list_unref /home/pel/sb/vim/src/list.c:151
    #1 0x5648415aa321 in clear_tv /home/pel/sb/vim/src/eval.c:7192
    #2 0x564841a51fc3 in get_func_tv /home/pel/sb/vim/src/userfunc.c:469
    #3 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #4 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #5 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #6 0x564841a548b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #7 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #8 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #9 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #10 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #11 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #12 0x5648415b09e6 in ex_execute /home/pel/sb/vim/src/eval.c:8536
    #13 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #14 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #15 0x564841a548b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #16 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #17 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #18 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #19 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #20 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #21 0x56484163c025 in do_source /home/pel/sb/vim/src/ex_cmds2.c:4615
    #22 0x56484163a81b in cmd_source /home/pel/sb/vim/src/ex_cmds2.c:4228
    #23 0x56484163a5f6 in ex_source /home/pel/sb/vim/src/ex_cmds2.c:4203
    #24 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #25 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #26 0x564841640f59 in do_cmdline_cmd /home/pel/sb/vim/src/ex_docmd.c:634
    #27 0x564841b2cf91 in exe_commands /home/pel/sb/vim/src/main.c:2959
    #28 0x564841b26161 in vim_main2 /home/pel/sb/vim/src/main.c:809
    #29 0x564841b258a4 in main /home/pel/sb/vim/src/main.c:441
    #30 0x7efd7b762b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #31 0x56484150b3f9 in _start (/home/pel/sb/vim/src/vim+0x1023f9)

0x6070000038c8 is located 56 bytes inside of 80-byte region [0x607000003890,0x6070000038e0)
freed by thread T0 here:
    #0 0x7efd7dca17b8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xde7b8)
    #1 0x56484178eefc in vim_free /home/pel/sb/vim/src/misc2.c:1817
    #2 0x564841734037 in list_free_list /home/pel/sb/vim/src/list.c:208
    #3 0x564841734110 in list_free_items /home/pel/sb/vim/src/list.c:225
    #4 0x5648415a1c1e in free_unref_items /home/pel/sb/vim/src/eval.c:5503
    #5 0x5648415a1b19 in garbage_collect /home/pel/sb/vim/src/eval.c:5449
    #6 0x564841a52d8e in cleanup_function_call /home/pel/sb/vim/src/userfunc.c:685
    #7 0x564841a5558d in call_user_func /home/pel/sb/vim/src/userfunc.c:1066
    #8 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #9 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #10 0x56484159bf24 in eval7 /home/pel/sb/vim/src/eval.c:4382
    #11 0x56484159a462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #12 0x564841599615 in eval5 /home/pel/sb/vim/src/eval.c:3781
    #13 0x564841598d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #14 0x564841598905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #15 0x564841598463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #16 0x564841597eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #17 0x564841a51c9f in get_func_tv /home/pel/sb/vim/src/userfunc.c:425
    #18 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #19 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #20 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #21 0x564841a548b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #22 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #23 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #24 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #25 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #26 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #27 0x5648415b09e6 in ex_execute /home/pel/sb/vim/src/eval.c:8536
    #28 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #29 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033

previously allocated by thread T0 here:
    #0 0x7efd7dca1b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x56484178c2c4 in lalloc /home/pel/sb/vim/src/misc2.c:944
    #2 0x56484178c137 in alloc_clear /home/pel/sb/vim/src/misc2.c:853
    #3 0x564841733976 in list_alloc /home/pel/sb/vim/src/list.c:75
    #4 0x5648417367c2 in get_list_tv /home/pel/sb/vim/src/list.c:890
    #5 0x56484159b814 in eval7 /home/pel/sb/vim/src/eval.c:4292
    #6 0x56484159a462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #7 0x564841599615 in eval5 /home/pel/sb/vim/src/eval.c:3781
    #8 0x564841598d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #9 0x564841598905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #10 0x564841598463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #11 0x564841597eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #12 0x564841a51c9f in get_func_tv /home/pel/sb/vim/src/userfunc.c:425
    #13 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #14 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #15 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #16 0x564841a548b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #17 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #18 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #19 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #20 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #21 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #22 0x5648415b09e6 in ex_execute /home/pel/sb/vim/src/eval.c:8536
    #23 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #24 0x5648416428b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #25 0x564841a548b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #26 0x564841a573a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #27 0x564841a51f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #28 0x564841a621e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #29 0x56484164b39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520

SUMMARY: AddressSanitizer: heap-use-after-free /home/pel/sb/vim/src/list.c:151 in list_unref
Shadow bytes around the buggy address:
  0x0c0e7fff86c0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0e7fff86d0: fd fd fd fd fd fa fa fa fa fa fd fd fd fd fd fd
  0x0c0e7fff86e0: fd fd fd fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0e7fff86f0: fd fd fa fa fa fa 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff8700: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 fa fa
=>0x0c0e7fff8710: fa fa fd fd fd fd fd fd fd[fd]fd fd fa fa fa fa
  0x0c0e7fff8720: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fd fd
  0x0c0e7fff8730: fd fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa
  0x0c0e7fff8740: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8750: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8760: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==10998==ABORTING
make[1]: *** [test_lambda.res] Error 1
make: *** [test_lambda] Error 2
```

test87 also crashes with an asan error:
```
==24205==ERROR: AddressSanitizer: heap-use-after-free on address 0x607000032094 at pc 0x561c15944eb5 bp 0x7fff52bf1b60 sp 0x7fff52bf1b50
READ of size 4 at 0x607000032094 thread T0
    #0 0x561c15944eb4 in echo_string_core /home/pel/sb/vim/src/eval.c:5915
    #1 0x561c159457da in tv2string /home/pel/sb/vim/src/eval.c:6024
    #2 0x561c15995d25 in f_string /home/pel/sb/vim/src/evalfunc.c:12703
    #3 0x561c1595db00 in call_internal_func /home/pel/sb/vim/src/evalfunc.c:1117
    #4 0x561c15df84fd in call_func /home/pel/sb/vim/src/userfunc.c:1520
    #5 0x561c15df2f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #6 0x561c1593cf24 in eval7 /home/pel/sb/vim/src/eval.c:4382
    #7 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #8 0x561c1593a824 in eval5 /home/pel/sb/vim/src/eval.c:3818
    #9 0x561c15939d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #10 0x561c15939905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #11 0x561c15939463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #12 0x561c15938eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #13 0x561c15938bf1 in eval0 /home/pel/sb/vim/src/eval.c:3403
    #14 0x561c15929e51 in eval_to_string /home/pel/sb/vim/src/eval.c:848
    #15 0x561c15ba12c4 in get_expr_line /home/pel/sb/vim/src/ops.c:840
    #16 0x561c15ba3344 in get_spec_reg /home/pel/sb/vim/src/ops.c:1528
    #17 0x561c15bb2153 in do_put /home/pel/sb/vim/src/ops.c:3444
    #18 0x561c15a125cf in ex_put /home/pel/sb/vim/src/ex_docmd.c:9485
    #19 0x561c159ec39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #20 0x561c159e38b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #21 0x561c15df58b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #22 0x561c15df83a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #23 0x561c15df2f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #24 0x561c15e031e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #25 0x561c159ec39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #26 0x561c159e38b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #27 0x561c15df58b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #28 0x561c15df83a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #29 0x561c15df2f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #30 0x561c15e031e4 in ex_call /home/pel/sb/vim/src/userfunc.c:3184
    #31 0x561c159ec39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #32 0x561c159e38b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #33 0x561c15b7fefe in nv_colon /home/pel/sb/vim/src/normal.c:5326
    #34 0x561c15b64925 in normal_cmd /home/pel/sb/vim/src/normal.c:1117
    #35 0x561c15ec828f in main_loop /home/pel/sb/vim/src/main.c:1370
    #36 0x561c15ec73de in vim_main2 /home/pel/sb/vim/src/main.c:912
    #37 0x561c15ec68a4 in main /home/pel/sb/vim/src/main.c:441
    #38 0x7f19ab838b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #39 0x561c158ac3f9 in _start (/home/pel/sb/vim/src/vim+0x1023f9)

0x607000032094 is located 68 bytes inside of 80-byte region [0x607000032050,0x6070000320a0)
freed by thread T0 here:
    #0 0x7f19add777b8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xde7b8)
    #1 0x561c15b2fefc in vim_free /home/pel/sb/vim/src/misc2.c:1817
    #2 0x561c15ad5037 in list_free_list /home/pel/sb/vim/src/list.c:208
    #3 0x561c15ad5110 in list_free_items /home/pel/sb/vim/src/list.c:225
    #4 0x561c15942c1e in free_unref_items /home/pel/sb/vim/src/eval.c:5503
    #5 0x561c15942b19 in garbage_collect /home/pel/sb/vim/src/eval.c:5449
    #6 0x561c15df3d8e in cleanup_function_call /home/pel/sb/vim/src/userfunc.c:685
    #7 0x561c15df658d in call_user_func /home/pel/sb/vim/src/userfunc.c:1066
    #8 0x561c15df83a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #9 0x561c15df2f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #10 0x561c1594a82b in handle_subscript /home/pel/sb/vim/src/eval.c:7034
    #11 0x561c1593d202 in eval7 /home/pel/sb/vim/src/eval.c:4422
    #12 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #13 0x561c1593a615 in eval5 /home/pel/sb/vim/src/eval.c:3781
    #14 0x561c15939d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #15 0x561c15939905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #16 0x561c15939463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #17 0x561c15938eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #18 0x561c15df2c9f in get_func_tv /home/pel/sb/vim/src/userfunc.c:425
    #19 0x561c1593cf24 in eval7 /home/pel/sb/vim/src/eval.c:4382
    #20 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #21 0x561c1593a824 in eval5 /home/pel/sb/vim/src/eval.c:3818
    #22 0x561c15939d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #23 0x561c15939905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #24 0x561c15939463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #25 0x561c15938eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #26 0x561c15938bf1 in eval0 /home/pel/sb/vim/src/eval.c:3403
    #27 0x561c15929e51 in eval_to_string /home/pel/sb/vim/src/eval.c:848
    #28 0x561c15ba12c4 in get_expr_line /home/pel/sb/vim/src/ops.c:840
    #29 0x561c15ba3344 in get_spec_reg /home/pel/sb/vim/src/ops.c:1528

previously allocated by thread T0 here:
    #0 0x7f19add77b50 in __interceptor_malloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xdeb50)
    #1 0x561c15b2d2c4 in lalloc /home/pel/sb/vim/src/misc2.c:944
    #2 0x561c15b2d137 in alloc_clear /home/pel/sb/vim/src/misc2.c:853
    #3 0x561c15ad4976 in list_alloc /home/pel/sb/vim/src/list.c:75
    #4 0x561c15ad77c2 in get_list_tv /home/pel/sb/vim/src/list.c:890
    #5 0x561c1593c814 in eval7 /home/pel/sb/vim/src/eval.c:4292
    #6 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #7 0x561c1593a615 in eval5 /home/pel/sb/vim/src/eval.c:3781
    #8 0x561c15939d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #9 0x561c15939905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #10 0x561c15939463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #11 0x561c15938eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #12 0x561c15938bf1 in eval0 /home/pel/sb/vim/src/eval.c:3403
    #13 0x561c15e024f4 in ex_return /home/pel/sb/vim/src/userfunc.c:3059
    #14 0x561c159ec39e in do_one_cmd /home/pel/sb/vim/src/ex_docmd.c:2520
    #15 0x561c159e38b6 in do_cmdline /home/pel/sb/vim/src/ex_docmd.c:1033
    #16 0x561c15df58b5 in call_user_func /home/pel/sb/vim/src/userfunc.c:967
    #17 0x561c15df83a5 in call_func /home/pel/sb/vim/src/userfunc.c:1501
    #18 0x561c15df2f44 in get_func_tv /home/pel/sb/vim/src/userfunc.c:455
    #19 0x561c1594a82b in handle_subscript /home/pel/sb/vim/src/eval.c:7034
    #20 0x561c1593d202 in eval7 /home/pel/sb/vim/src/eval.c:4422
    #21 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985
    #22 0x561c1593a615 in eval5 /home/pel/sb/vim/src/eval.c:3781
    #23 0x561c15939d71 in eval4 /home/pel/sb/vim/src/eval.c:3665
    #24 0x561c15939905 in eval3 /home/pel/sb/vim/src/eval.c:3585
    #25 0x561c15939463 in eval2 /home/pel/sb/vim/src/eval.c:3517
    #26 0x561c15938eec in eval1 /home/pel/sb/vim/src/eval.c:3445
    #27 0x561c15df2c9f in get_func_tv /home/pel/sb/vim/src/userfunc.c:425
    #28 0x561c1593cf24 in eval7 /home/pel/sb/vim/src/eval.c:4382
    #29 0x561c1593b462 in eval6 /home/pel/sb/vim/src/eval.c:3985

SUMMARY: AddressSanitizer: heap-use-after-free /home/pel/sb/vim/src/eval.c:5915 in echo_string_core
Shadow bytes around the buggy address:
  0x0c0e7fffe3c0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fa fa
  0x0c0e7fffe3d0: fa fa fd fd fd fd fd fd fd fd fd fa fa fa fa fa
  0x0c0e7fffe3e0: fd fd fd fd fd fd fd fd fd fa fa fa fa fa fd fd
  0x0c0e7fffe3f0: fd fd fd fd fd fd fd fd fa fa fa fa 00 00 00 00
  0x0c0e7fffe400: 00 00 00 00 00 00 fa fa fa fa fd fd fd fd fd fd
=>0x0c0e7fffe410: fd fd[fd]fd fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fffe420: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fffe430: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fffe440: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fffe450: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fffe460: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==24205==ABORTING
Makefile:97: recipe for target 'test87.out' failed
make: *** [test87.out] Error 1
```

All other tests pass.

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -442,7 +442,8 @@ void eval_clear(void)
   // unreferenced lists and dicts
   (void)garbage_collect(false);
 
-  // functions
+  // functions need to be freed before gargabe collecting, otherwise local
+  // variables might be freed twice.
   free_all_functions();
 }
 

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -439,12 +439,12 @@ void eval_clear(void)
     xfree(SCRIPT_SV(i));
   ga_clear(&ga_scripts);
 
-  // unreferenced lists and dicts
-  (void)garbage_collect(false);
-
   // functions need to be freed before gargabe collecting, otherwise local
   // variables might be freed twice.
   free_all_functions();
+
+  // unreferenced lists and dicts
+  (void)garbage_collect(false);
 }
 
 #endif

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -439,12 +439,11 @@ void eval_clear(void)
     xfree(SCRIPT_SV(i));
   ga_clear(&ga_scripts);
 
-  // functions need to be freed before gargabe collecting, otherwise local
-  // variables might be freed twice.
-  free_all_functions();
-
   // unreferenced lists and dicts
   (void)garbage_collect(false);
+
+  // functions not garbage collected
+  free_all_functions();
 }
 
 #endif

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -871,17 +871,19 @@ char_u *eval_to_string(char_u *arg, char_u **nextcmd, int convert)
 char_u *eval_to_string_safe(char_u *arg, char_u **nextcmd, int use_sandbox)
 {
   char_u      *retval;
-  void        *save_funccalp;
+  funccal_entry_T funccal_entry;
 
-  save_funccalp = save_funccal();
-  if (use_sandbox)
-    ++sandbox;
-  ++textlock;
-  retval = eval_to_string(arg, nextcmd, FALSE);
-  if (use_sandbox)
-    --sandbox;
-  --textlock;
-  restore_funccal(save_funccalp);
+  save_funccal(&funccal_entry);
+  if (use_sandbox) {
+    sandbox++;
+  }
+  textlock++;
+  retval = eval_to_string(arg, nextcmd, false);
+  if (use_sandbox) {
+    sandbox--;
+  }
+  textlock--;
+  restore_funccal();
   return retval;
 }
 
@@ -9747,9 +9749,11 @@ const void *var_shada_iter(const void *const iter, const char **const name,
 
 void var_set_global(const char *const name, typval_T vartv)
 {
-  funccall_T *const saved_funccal = (funccall_T *)save_funccal();
+  funccal_entry_T funccall_entry;
+
+  save_funccal(&funccall_entry);
   set_var(name, strlen(name), &vartv, false);
-  restore_funccal(saved_funccal);
+  restore_funccal();
 }
 
 int store_session_globals(FILE *fd)
@@ -10305,8 +10309,10 @@ typval_T eval_call_provider(char *provider, char *method, list_T *arguments)
     .autocmd_fname = autocmd_fname,
     .autocmd_match = autocmd_match,
     .autocmd_bufnr = autocmd_bufnr,
-    .funccalp = save_funccal()
+    .funccalp = (void *)get_current_funccal()
   };
+  funccal_entry_T funccal_entry;
+  save_funccal(&funccal_entry);
   provider_call_nesting++;
 
   typval_T argvars[3] = {
@@ -10333,7 +10339,7 @@ typval_T eval_call_provider(char *provider, char *method, list_T *arguments)
 
   tv_list_unref(arguments);
   // Restore caller scope information
-  restore_funccal(provider_caller_scope.funccalp);
+  restore_funccal();
   provider_caller_scope = saved_provider_caller_scope;
   provider_call_nesting--;
   assert(provider_call_nesting >= 0);

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -5026,7 +5026,7 @@ bool garbage_collect(bool testing)
 
     // 3. Check if any funccal can be freed now.
     //    This may call us back recursively.
-    did_free = did_free || free_unref_funccal(copyID, testing);
+    did_free = free_unref_funccal(copyID, testing) || did_free;
   } else if (p_verbose > 0) {
     verb_msg(_(
         "Not enough memory to set references, garbage collection aborted!"));

--- a/src/nvim/eval/funcs.c
+++ b/src/nvim/eval/funcs.c
@@ -7219,7 +7219,7 @@ static void f_rpcrequest(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   uint8_t *save_sourcing_name, *save_autocmd_fname, *save_autocmd_match;
   linenr_T save_sourcing_lnum;
   int save_autocmd_bufnr;
-  void *save_funccalp;
+  funccal_entry_T funccal_entry;
 
   if (l_provider_call_nesting) {
     // If this is called from a provider function, restore the scope
@@ -7230,7 +7230,7 @@ static void f_rpcrequest(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     save_autocmd_fname = autocmd_fname;
     save_autocmd_match = autocmd_match;
     save_autocmd_bufnr = autocmd_bufnr;
-    save_funccalp = save_funccal();
+    save_funccal(&funccal_entry);
 
     current_sctx = provider_caller_scope.script_ctx;
     sourcing_name = provider_caller_scope.sourcing_name;
@@ -7238,7 +7238,7 @@ static void f_rpcrequest(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     autocmd_fname = provider_caller_scope.autocmd_fname;
     autocmd_match = provider_caller_scope.autocmd_match;
     autocmd_bufnr = provider_caller_scope.autocmd_bufnr;
-    restore_funccal(provider_caller_scope.funccalp);
+    set_current_funccal((funccall_T *)(provider_caller_scope.funccalp));
   }
 
 
@@ -7256,7 +7256,7 @@ static void f_rpcrequest(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     autocmd_fname = save_autocmd_fname;
     autocmd_match = save_autocmd_match;
     autocmd_bufnr = save_autocmd_bufnr;
-    restore_funccal(save_funccalp);
+    restore_funccal();
   }
 
   if (ERROR_SET(&err)) {

--- a/src/nvim/eval/typval_encode.c.h
+++ b/src/nvim/eval/typval_encode.c.h
@@ -607,7 +607,7 @@ _convert_one_value_regular_dict: {}
                                              kMPConvDict);
       TYPVAL_ENCODE_CONV_DICT_START(tv, tv->vval.v_dict,
                                     tv->vval.v_dict->dv_hashtab.ht_used);
-      assert(saved_copyID != copyID && saved_copyID != copyID - 1);
+      assert(saved_copyID != copyID);
       _mp_push(*mpstack, ((MPConvStackVal) {
         .tv = tv,
         .type = kMPConvDict,

--- a/src/nvim/eval/userfunc.c
+++ b/src/nvim/eval/userfunc.c
@@ -3356,9 +3356,18 @@ bool set_ref_in_call_stack(int copyID)
   bool abort = false;
 
   for (funccall_T *fc = current_funccal; fc != NULL; fc = fc->caller) {
-    abort = abort || set_ref_in_ht(&fc->l_vars.dv_hashtab, copyID, NULL);
-    abort = abort || set_ref_in_ht(&fc->l_avars.dv_hashtab, copyID, NULL);
+    abort = abort || set_ref_in_funccal(fc, copyID);
   }
+
+  // Also go through the funccal_stack.
+  for (funccal_entry_T *entry = funccal_stack; entry != NULL;
+       entry = entry->next) {
+    for (funccall_T *fc = entry->top_funccal; !abort && fc != NULL;
+         fc = fc->caller) {
+      abort = abort || set_ref_in_funccal(fc, copyID);
+    }
+  }
+
   return abort;
 }
 

--- a/src/nvim/eval/userfunc.c
+++ b/src/nvim/eval/userfunc.c
@@ -43,11 +43,11 @@ hashtab_T func_hashtab;
 static garray_T funcargs = GA_EMPTY_INIT_VALUE;
 
 // pointer to funccal for currently active function
-funccall_T *current_funccal = NULL;
+static funccall_T *current_funccal = NULL;
 
 // Pointer to list of previously used funccal, still around because some
 // item in it is still being used.
-funccall_T *previous_funccal = NULL;
+static funccall_T *previous_funccal = NULL;
 
 static char *e_funcexts = N_(
     "E122: Function %s already exists, add ! to replace it");
@@ -541,14 +541,8 @@ static void add_nr_var(dict_T *dp, dictitem_T *v, char *name, varnumber_T nr)
   v->di_tv.vval.v_number = nr;
 }
 
-/*
- * Free "fc" and what it contains.
- */
-static void
-free_funccal(
-    funccall_T *fc,
-    int free_val              // a: vars were allocated
-)
+// Free "fc"
+static void free_funccal(funccall_T *fc)
 {
   for (int i = 0; i < fc->fc_funcs.ga_len; i++) {
     ufunc_T *fp = ((ufunc_T **)(fc->fc_funcs.ga_data))[i];
@@ -563,34 +557,74 @@ free_funccal(
   }
   ga_clear(&fc->fc_funcs);
 
-  // The a: variables typevals may not have been allocated, only free the
-  // allocated variables.
-  vars_clear_ext(&fc->l_avars.dv_hashtab, free_val);
+  func_ptr_unref(fc->func);
+  xfree(fc);
+}
 
+// Free "fc" and what it contains.
+// Can be called only when "fc" is kept beyond the period of it called,
+// i.e. after cleanup_function_call(fc).
+static void free_funccal_contents(funccall_T *fc)
+{
   // Free all l: variables.
   vars_clear(&fc->l_vars.dv_hashtab);
 
-  // Free the a:000 variables if they were allocated.
-  if (free_val) {
-    TV_LIST_ITER(&fc->l_varlist, li, {
-      tv_clear(TV_LIST_ITEM_TV(li));
-    });
-  }
+  // Free all a: variables.
+  vars_clear(&fc->l_avars.dv_hashtab);
 
-  func_ptr_unref(fc->func);
-  xfree(fc);
+  // Free the a:000 variables.
+  TV_LIST_ITER(&fc->l_varlist, li, {
+    tv_clear(TV_LIST_ITEM_TV(li));
+  });
+
+  free_funccal(fc);
 }
 
 /// Handle the last part of returning from a function: free the local hashtable.
 /// Unless it is still in use by a closure.
 static void cleanup_function_call(funccall_T *fc)
 {
+  bool may_free_fc = fc->fc_refcount <= 0;
+  bool free_fc = true;
+
   current_funccal = fc->caller;
 
-  // If the a:000 list and the l: and a: dicts are not referenced and there
-  // is no closure using it, we can free the funccall_T and what's in it.
-  if (!fc_referenced(fc)) {
-    free_funccal(fc, false);
+  // Free all l: variables if not referred.
+  if (may_free_fc && fc->l_vars.dv_refcount == DO_NOT_FREE_CNT) {
+    vars_clear(&fc->l_vars.dv_hashtab);
+  } else {
+    free_fc = false;
+  }
+
+  // If the a:000 list and the l: and a: dicts are not referenced and
+  // there is no closure using it, we can free the funccall_T and what's
+  // in it.
+  if (may_free_fc && fc->l_avars.dv_refcount == DO_NOT_FREE_CNT) {
+    vars_clear_ext(&fc->l_avars.dv_hashtab, false);
+  } else {
+    free_fc = false;
+
+    // Make a copy of the a: variables, since we didn't do that above.
+    TV_DICT_ITER(&fc->l_avars, di, {
+      tv_copy(&di->di_tv, &di->di_tv);
+    });
+  }
+
+  if (may_free_fc && fc->l_varlist.lv_refcount   // NOLINT(runtime/deprecated)
+      == DO_NOT_FREE_CNT) {
+    fc->l_varlist.lv_first = NULL;  // NOLINT(runtime/deprecated)
+
+  } else {
+    free_fc = false;
+
+    // Make a copy of the a:000 items, since we didn't do that above.
+    TV_LIST_ITER(&fc->l_varlist, li, {
+      tv_copy(TV_LIST_ITEM_TV(li), TV_LIST_ITEM_TV(li));
+    });
+  }
+
+  if (free_fc) {
+    free_funccal(fc);
   } else {
     static int made_copy = 0;
 
@@ -600,19 +634,12 @@ static void cleanup_function_call(funccall_T *fc)
     fc->caller = previous_funccal;
     previous_funccal = fc;
 
-    // Make a copy of the a: variables, since we didn't do that above.
-    TV_DICT_ITER(&fc->l_avars, di, {
-      tv_copy(&di->di_tv, &di->di_tv);
-    });
-
-    // Make a copy of the a:000 items, since we didn't do that above.
-    TV_LIST_ITER(&fc->l_varlist, li, {
-      tv_copy(TV_LIST_ITEM_TV(li), TV_LIST_ITEM_TV(li));
-    });
-
-    if (++made_copy == 10000) {
-      // We have made a lot of copies.  This can happen when
-      // repetitively calling a function that creates a reference to
+    if (want_garbage_collect) {
+      // If garbage collector is ready, clear count.
+      made_copy = 0;
+    } else if (++made_copy >= (int)((4096 * 1024) / sizeof(*fc))) {
+      // We have made a lot of copies, worth 4 Mbyte.  This can happen
+      // when repetitively calling a function that creates a reference to
       // itself somehow.  Call the garbage collector soon to avoid using
       // too much memory.
       made_copy = 0;
@@ -639,7 +666,7 @@ static void funccal_unref(funccall_T *fc, ufunc_T *fp, bool force)
     for (pfc = &previous_funccal; *pfc != NULL; pfc = &(*pfc)->caller) {
       if (fc == *pfc) {
         *pfc = fc->caller;
-        free_funccal(fc, true);
+        free_funccal_contents(fc);
         return;
       }
     }
@@ -766,7 +793,7 @@ void call_user_func(ufunc_T *fp, int argcount, typval_T *argvars,
   // check for CTRL-C hit
   line_breakcheck();
   // prepare the funccall_T structure
-  fc = xmalloc(sizeof(funccall_T));
+  fc = xcalloc(1, sizeof(funccall_T));
   fc->caller = current_funccal;
   current_funccal = fc;
   fc->func = fp;
@@ -881,9 +908,11 @@ void call_user_func(ufunc_T *fp, int argcount, typval_T *argvars,
     }
 
     if (ai >= 0 && ai < MAX_FUNC_ARGS) {
-      tv_list_append(&fc->l_varlist, &fc->l_listitems[ai]);
-      *TV_LIST_ITEM_TV(&fc->l_listitems[ai]) = argvars[i];
-      TV_LIST_ITEM_TV(&fc->l_listitems[ai])->v_lock = VAR_FIXED;
+      listitem_T *li = &fc->l_listitems[ai];
+
+      *TV_LIST_ITEM_TV(li) = argvars[i];
+      TV_LIST_ITEM_TV(li)->v_lock =  VAR_FIXED;
+      tv_list_append(&fc->l_varlist, li);
     }
   }
 
@@ -3134,7 +3163,7 @@ bool free_unref_funccal(int copyID, int testing)
     if (can_free_funccal(*pfc, copyID)) {
       funccall_T *fc = *pfc;
       *pfc = fc->caller;
-      free_funccal(fc, true);
+      free_funccal_contents(fc);
       did_free = true;
       did_free_funccal = true;
     } else {

--- a/src/nvim/eval/userfunc.c
+++ b/src/nvim/eval/userfunc.c
@@ -1106,26 +1106,36 @@ static bool func_name_refcount(char_u *name)
   return isdigit(*name) || *name == '<';
 }
 
-/*
- * Save the current function call pointer, and set it to NULL.
- * Used when executing autocommands and for ":source".
- */
-void *save_funccal(void)
-{
-  funccall_T *fc = current_funccal;
+static funccal_entry_T *funccal_stack = NULL;
 
+// Save the current function call pointer, and set it to NULL.
+// Used when executing autocommands and for ":source".
+void save_funccal(funccal_entry_T *entry)
+{
+  entry->top_funccal = current_funccal;
+  entry->next = funccal_stack;
+  funccal_stack = entry;
   current_funccal = NULL;
-  return (void *)fc;
 }
 
-void restore_funccal(void *vfc)
+void restore_funccal(void)
 {
-  current_funccal = (funccall_T *)vfc;
+  if (funccal_stack == NULL) {
+    IEMSG("INTERNAL: restore_funccal()");
+  } else {
+    current_funccal = funccal_stack->top_funccal;
+    funccal_stack = funccal_stack->next;
+  }
 }
 
 funccall_T *get_current_funccal(void)
 {
   return current_funccal;
+}
+
+void set_current_funccal(funccall_T *fc)
+{
+  current_funccal = fc;
 }
 
 #if defined(EXITFREE)
@@ -1137,10 +1147,13 @@ void free_all_functions(void)
   uint64_t todo = 1;
   uint64_t used;
 
-  // Clean up the call stack.
+  // Clean up the current_funccal chain and the funccal stack.
   while (current_funccal != NULL) {
     tv_clear(current_funccal->rettv);
     cleanup_function_call(current_funccal);
+    if (current_funccal == NULL && funccal_stack != NULL) {
+      restore_funccal();
+    }
   }
 
   // First clear what the functions contain. Since this may lower the

--- a/src/nvim/eval/userfunc.h
+++ b/src/nvim/eval/userfunc.h
@@ -11,6 +11,12 @@ typedef struct {
   dictitem_T  *fd_di;  ///< Dictionary item used.
 } funcdict_T;
 
+typedef struct funccal_entry funccal_entry_T;
+struct funccal_entry {
+  void *top_funccal;
+  funccal_entry_T *next;
+};
+
 /// errors for when calling a function
 typedef enum {
   ERROR_UNKNOWN = 0,

--- a/src/nvim/ex_cmds2.c
+++ b/src/nvim/ex_cmds2.c
@@ -3106,7 +3106,6 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
   int retval = FAIL;
   static scid_T last_current_SID = 0;
   static int last_current_SID_seq = 0;
-  void                    *save_funccalp;
   int save_debug_break_level = debug_break_level;
   scriptitem_T            *si = NULL;
   proftime_T wait_start;
@@ -3227,7 +3226,8 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
 
   // Don't use local function variables, if called from a function.
   // Also starts profiling timer for nested script.
-  save_funccalp = save_funccal();
+  funccal_entry_T funccalp_entry;
+  save_funccal(&funccalp_entry);
 
   // Check if this script was sourced before to finds its SID.
   // If it's new, generate a new SID.
@@ -3352,7 +3352,7 @@ int do_source(char_u *fname, int check_other, int is_vimrc)
   }
 
   current_sctx = save_current_sctx;
-  restore_funccal(save_funccalp);
+  restore_funccal();
   if (l_do_profiling == PROF_YES) {
     prof_child_exit(&wait_start);    // leaving a child now
   }

--- a/src/nvim/fileio.c
+++ b/src/nvim/fileio.c
@@ -6736,7 +6736,6 @@ static bool apply_autocmds_group(event_T event, char_u *fname, char_u *fname_io,
   static int nesting = 0;
   AutoPatCmd patcmd;
   AutoPat     *ap;
-  void        *save_funccalp;
   char_u      *save_cmdarg;
   long save_cmdbang;
   static int filechangeshell_busy = FALSE;
@@ -6926,8 +6925,9 @@ static bool apply_autocmds_group(event_T event, char_u *fname, char_u *fname_io,
   if (do_profiling == PROF_YES)
     prof_child_enter(&wait_time);     /* doesn't count for the caller itself */
 
-  /* Don't use local function variables, if called from a function */
-  save_funccalp = save_funccal();
+  // Don't use local function variables, if called from a function.
+  funccal_entry_T funccal_entry;
+  save_funccal(&funccal_entry);
 
   /*
    * When starting to execute autocommands, save the search patterns.
@@ -7016,9 +7016,10 @@ static bool apply_autocmds_group(event_T event, char_u *fname, char_u *fname_io,
   autocmd_bufnr = save_autocmd_bufnr;
   autocmd_match = save_autocmd_match;
   current_sctx = save_current_sctx;
-  restore_funccal(save_funccalp);
-  if (do_profiling == PROF_YES)
+  restore_funccal();
+  if (do_profiling == PROF_YES) {
     prof_child_exit(&wait_time);
+  }
   KeyTyped = save_KeyTyped;
   xfree(fname);
   xfree(sfname);

--- a/src/nvim/testdir/runtest.vim
+++ b/src/nvim/testdir/runtest.vim
@@ -44,6 +44,10 @@ if &lines < 24 || &columns < 80
   qa!
 endif
 
+if has('reltime')
+  let s:start_time = reltime()
+endif
+
 " Common with all tests on all systems.
 source setup.vim
 
@@ -100,6 +104,9 @@ endfunc
 
 func RunTheTest(test)
   echo 'Executing ' . a:test
+  if has('reltime')
+    let func_start = reltime()
+  endif
 
   " Avoid stopping at the "hit enter" prompt
   set nomore
@@ -124,7 +131,11 @@ func RunTheTest(test)
     endtry
   endif
 
-  call add(s:messages, 'Executing ' . a:test)
+  let message = 'Executed ' . a:test
+  if has('reltime')
+    let message ..= ' in ' .. reltimestr(reltime(func_start)) .. ' seconds'
+  endif
+  call add(s:messages, message)
   let s:done += 1
 
   if a:test =~ 'Test_nocatch_'
@@ -229,6 +240,9 @@ func FinishTesting()
     let message = 'NO tests executed'
   else
     let message = 'Executed ' . s:done . (s:done > 1 ? ' tests' : ' test')
+  endif
+  if has('reltime')
+    let message ..= ' in ' .. reltimestr(reltime(s:start_time)) .. ' seconds'
   endif
   echo message
   call add(s:messages, message)

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -21,9 +21,7 @@ local monitor_memory_usage = {
   op = function(self)
     retry(nil, 10000, function()
       local val = self.memory_usage(self)
-      if self.min > val then
-        self.min = val
-      elseif self.max < val then
+      if self.max < val then
         self.max = val
       end
       table.insert(self.hist, val)
@@ -45,7 +43,6 @@ local monitor_memory_usage = {
   monitor_memory_usage = function(self, pid)
     local obj = {
       pid = pid,
-      min = 0,
       max = 0,
       last = 0,
       hist = {},
@@ -106,8 +103,13 @@ describe('memory usage', function()
     -- Estimate the limit of max usage as 2x initial usage.
     check_result({before=before, after=after}, pcall(ok, before.last < after.max))
     check_result({before=before, after=after}, pcall(ok, before.last * 2 > after.max))
-    -- In this case, garbase collecting is not needed.
-    check_result({before=before, after=after}, pcall(eq, after.last, after.max))
+    -- In this case, garbage collecting is not needed.
+    -- The value might fluctuate
+    -- a bit, allow for 3% tolerance.
+    local lower = after.last * 97 / 100
+    local upper = after.last * 103 / 100
+    check_result({before=before, after=after}, pcall(ok, lower < after.max))
+    check_result({before=before, after=after}, pcall(ok, after.max < upper))
   end)
 
   --[[
@@ -143,11 +145,13 @@ describe('memory usage', function()
       feed_command('so '..fname)
     end
     local last = monitor_memory_usage(pid)
-    -- The usage may be a bit less than the last value
+    -- The usage may be a bit less than the last value, use 80%.
+    -- Allow for 1% tolerance at the upper limit.
     local lower = before.last * 8 / 10
+    local upper = (after.max + (after.last - before.last)) * 101 / 100
     check_result({before=before, after=after, last=last},
                  pcall(ok, lower < last.last))
     check_result({before=before, after=after, last=last},
-                 pcall(ok, last.last < after.max + (after.last - before.last)))
+                 pcall(ok, last.last < upper))
   end)
 end)

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -143,8 +143,10 @@ describe('memory usage', function()
       feed_command('so '..fname)
     end
     local last = monitor_memory_usage(pid)
+    -- The usage may be a bit less than the last value
+    local lower = before.last * 8 / 10
     check_result({before=before, after=after, last=last},
-                 pcall(ok, before.last < last.last))
+                 pcall(ok, lower < last.last))
     check_result({before=before, after=after, last=last},
                  pcall(ok, last.last < after.max + (after.last - before.last)))
   end)

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -1,0 +1,151 @@
+local helpers = require('test.functional.helpers')(after_each)
+local clear = helpers.clear
+local eval = helpers.eval
+local eq = helpers.eq
+local feed_command = helpers.feed_command
+local iswin = helpers.iswin
+local retry = helpers.retry
+local ok = helpers.ok
+local source = helpers.source
+
+local monitor_memory_usage = {
+  memory_usage = function(self)
+    local handle
+    if iswin() then
+      handle = io.popen('wmic process where processid=' ..self.pid..' get WorkingSetSize')
+    else
+      handle = io.popen('ps -o rss= -p '..self.pid)
+    end
+    return tonumber(handle:read('*a'):match('%d+'))
+  end,
+  op = function(self)
+    retry(nil, 10000, function()
+      local val = self.memory_usage(self)
+      if self.min > val then
+        self.min = val
+      elseif self.max < val then
+        self.max = val
+      end
+      table.insert(self.hist, val)
+      ok(#self.hist > 20)
+      local result = {}
+      for key,value in ipairs(self.hist) do
+        if value ~= self.hist[key + 1] then
+          table.insert(result, value)
+        end
+      end
+      table.remove(self.hist, 1)
+      self.last = self.hist[#self.hist]
+      eq(#result, 1)
+    end)
+  end,
+  dump = function(self)
+    return 'max: '..self.max ..', last: '..self.last
+  end,
+  monitor_memory_usage = function(self, pid)
+    local obj = {
+      pid = pid,
+      min = 0,
+      max = 0,
+      last = 0,
+      hist = {},
+    }
+    setmetatable(obj, { __index = self })
+    obj:op()
+    return obj
+  end
+}
+setmetatable(monitor_memory_usage,
+{__call = function(self, pid)
+  return monitor_memory_usage.monitor_memory_usage(self, pid)
+end})
+
+describe('memory usage', function()
+  local function check_result(tbl, status, result)
+    if not status then
+      print('')
+      for key, val in pairs(tbl) do
+        print(key, val:dump())
+      end
+      error(result)
+    end
+  end
+
+  local function isasan()
+    local version = eval('execute("version")')
+    return version:match('-fsanitize=[a-z,]*address')
+  end
+
+  before_each(clear)
+
+  --[[
+  Case: if a local variable captures a:000, funccall object will be free
+  just after it finishes.
+  ]]--
+  it('function capture vargs', function()
+    if isasan() then
+      pending('ASAN build is difficult to estimate memory usage')
+    end
+    if iswin() and eval("executable('wmic')")  == 0 then
+      pending('missing "wmic" command')
+    elseif eval("executable('ps')")  == 0 then
+      pending('missing "ps" command')
+    end
+
+    local pid = eval('getpid()')
+    local before = monitor_memory_usage(pid)
+    source([[
+      func s:f(...)
+        let x = a:000
+      endfunc
+      for _ in range(10000)
+        call s:f(0)
+      endfor
+    ]])
+    local after = monitor_memory_usage(pid)
+    -- Estimate the limit of max usage as 2x initial usage.
+    check_result({before=before, after=after}, pcall(ok, before.last < after.max))
+    check_result({before=before, after=after}, pcall(ok, before.last * 2 > after.max))
+    -- In this case, garbase collecting is not needed.
+    check_result({before=before, after=after}, pcall(eq, after.last, after.max))
+  end)
+
+  --[[
+  Case: if a local variable captures l: dict, funccall object will not be
+  free until garbage collector runs, but after that memory usage doesn't
+  increase so much even when rerun Xtest.vim since system memory caches.
+  ]]--
+  it('function capture lvars', function()
+    if isasan() then
+      pending('ASAN build is difficult to estimate memory usage')
+    end
+    if iswin() and eval("executable('wmic')")  == 0 then
+      pending('missing "wmic" command')
+    elseif eval("executable('ps')")  == 0 then
+      pending('missing "ps" command')
+    end
+
+    local pid = eval('getpid()')
+    local before = monitor_memory_usage(pid)
+    local fname = source([[
+      if !exists('s:defined_func')
+        func s:f()
+          let x = l:
+        endfunc
+      endif
+      let s:defined_func = 1
+      for _ in range(10000)
+        call s:f()
+      endfor
+    ]])
+    local after = monitor_memory_usage(pid)
+    for _ = 1, 3 do
+      feed_command('so '..fname)
+    end
+    local last = monitor_memory_usage(pid)
+    check_result({before=before, after=after, last=last},
+                 pcall(ok, before.last < last.last))
+    check_result({before=before, after=after, last=last},
+                 pcall(ok, last.last < after.max + (after.last - before.last)))
+  end)
+end)

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -101,11 +101,13 @@ describe('memory usage', function()
     ]])
     local after = monitor_memory_usage(pid)
     -- Estimate the limit of max usage as 2x initial usage.
-    check_result({before=before, after=after}, pcall(ok, before.last < after.max))
-    check_result({before=before, after=after}, pcall(ok, before.last * 2 > after.max))
+    -- The lower limit can fluctuate a bit, use 98%.
+    check_result({before=before, after=after},
+                 pcall(ok, before.last * 98 / 100 < after.max))
+    check_result({before=before, after=after},
+                 pcall(ok, before.last * 2 > after.max))
     -- In this case, garbage collecting is not needed.
-    -- The value might fluctuate
-    -- a bit, allow for 3% tolerance.
+    -- The value might fluctuate a bit, allow for 3% tolerance.
     local lower = after.last * 97 / 100
     local upper = after.last * 103 / 100
     check_result({before=before, after=after}, pcall(ok, lower < after.max))

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -146,9 +146,9 @@ describe('memory usage', function()
     end
     local last = monitor_memory_usage(pid)
     -- The usage may be a bit less than the last value, use 80%.
-    -- Allow for 1% tolerance at the upper limit.
+    -- Allow for 3% tolerance at the upper limit.
     local lower = before.last * 8 / 10
-    local upper = (after.max + (after.last - before.last)) * 101 / 100
+    local upper = (after.max + (after.last - before.last)) * 103 / 100
     check_result({before=before, after=after, last=last},
                  pcall(ok, lower < last.last))
     check_result({before=before, after=after, last=last},

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -101,15 +101,16 @@ describe('memory usage', function()
     ]])
     local after = monitor_memory_usage(pid)
     -- Estimate the limit of max usage as 2x initial usage.
-    -- The lower limit can fluctuate a bit, use 98%.
+    -- The lower limit can fluctuate a bit, use 97%.
     check_result({before=before, after=after},
-                 pcall(ok, before.last * 98 / 100 < after.max))
+                 pcall(ok, before.last * 97 / 100 < after.max))
     check_result({before=before, after=after},
                  pcall(ok, before.last * 2 > after.max))
     -- In this case, garbage collecting is not needed.
-    -- The value might fluctuate a bit, allow for 3% tolerance.
+    -- The value might fluctuate a bit, allow for 3% tolerance below and 5% above.
+    -- Based on various test runs.
     local lower = after.last * 97 / 100
-    local upper = after.last * 103 / 100
+    local upper = after.last * 105 / 100
     check_result({before=before, after=after}, pcall(ok, lower < after.max))
     check_result({before=before, after=after}, pcall(ok, after.max < upper))
   end)

--- a/test/functional/legacy/memory_usage_spec.lua
+++ b/test/functional/legacy/memory_usage_spec.lua
@@ -146,9 +146,10 @@ describe('memory usage', function()
     end
     local last = monitor_memory_usage(pid)
     -- The usage may be a bit less than the last value, use 80%.
-    -- Allow for 3% tolerance at the upper limit.
+    -- Allow for 20% tolerance at the upper limit. That's very permissive, but
+    -- otherwise the test fails sometimes.
     local lower = before.last * 8 / 10
-    local upper = (after.max + (after.last - before.last)) * 103 / 100
+    local upper = (after.max + (after.last - before.last)) * 12 / 10
     check_result({before=before, after=after, last=last},
                  pcall(ok, lower < last.last))
     check_result({before=before, after=after, last=last},

